enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,246 @@
+so that consumers of serving certs and CA bundles can trivially support automated
+refresh. This is already supported by
+[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/pull/421)
+and the functionality could be separately extracted for operator reuse.
+
+## Proposal
+
+- Detect when the current CA certificate has passed 1/2 of its validity bounds
+  (i.e. 6 months has passed of a 1 year validity) and generate a new CA.
+  - The new CA should have the same `Subject.CommonName` as the current CA to
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Generate an intermediate CA certificate with the same public key as the current CA
+  but signed with the private key of the new CA.
+  - This intermediate certificate should be included in the the new CA
+    bundle along with the new CA's primary certificate, and serves to
+    bridge trust between an unrefreshed server (serving with a cert
+    generated from the previous CA) and a refreshed client (validating
+    trust with the new CA bundle).
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+  - In addition to ensuring regeneration when `AuthorityKeyId` and `SubjectKeyId`
+    are set, this comparison schema will also ensure regeneration of serving
+    certs without `SubjectKeyId` set (due to being generated by a CA without
+    `AuthorityKeyId` set) since an empty `SubjectKeyId` will not match a
+    populated `AuthorityKeyId`.
+
+- Provide an indication that rotation has taken place via an event.
+
+- Allow rotation to be forced when
+  `ServiceCA.Spec.ForceRotation.Reason` is set to a non-empty value
+  that does not match the value for the
+  `service.beta.openshift.io/forced-rotation-reason` annotation on the
+  signing secret.
+  - Setting an annotation on the signing secret ensures that a forced
```

Comment:
Is this to just ensure that the forced rotation does not collide with an automatic one? Would it be possible to force rotation with a new reason when a previously forced rotation is still happening?

Reply:
This strategy to prevent more than one forced rotation per reason was that suggested by @deads2k for etcd key rotation (via @sttts). It would not be possible to prompt 2 concurrent forced rotations (since the operator sync loop reconciles changes serially), but it would be possible to prompt a second rotation before clients of the service ca operator had fully refreshed. That window could be narrowed by ensuring hitless rotation as mentioned in the risks/mitigations section.

Reply:
Ok, I was thinking of the force rotation as covering all of the serving certs and client configmaps being refreshed.

Reply:
Fair enough. I'm trying to separate the updating of the signing secret from the updating of serving certs and injecting new ca bundles because the former is very much a constrained problem and the latter is not.

Reply:
Should `ServiceCA.Spec.ForceRotation.Reason` be `ServiceCA.Spec.ForceCARotation.Reason` ?

Reply:
Good catch! I've been using `ForceCARotation` in https://github.com/openshift/service-ca-operator/pull/77 but forgot to update the enhancement accordingly. 

Done.